plugins/event/sqs/sqs.go

```diff
@@ -0,0 +1,248 @@
+	}
+	options = append(options, awsutil.WithEnvironmentCredentials(true))
+	credConfig, err := awsutil.NewCredentialsConfig(options...)
+	if err != nil {
+		return nil, err
+	}
+	session, err := credConfig.GetSession()
+	if err != nil {
+		return nil, err
+	}
+	return sqs.New(session), nil
+}
+
+func (s *sqsBackend) Subscribe(_ context.Context, request *event.SubscribeRequest) error {
+	var sconfig sqsConfig
+	err := mapstructure.Decode(request.Config, &sconfig)
+	if err != nil {
+		return err
+	}
+	if sconfig.QueueName == "" && sconfig.QueueURL == "" {
+		return ErrQueueRequired
+	}
+	client, err := newClient(&sconfig)
+	if err != nil {
+		return err
+	}
+	var queueURL string
+	if sconfig.CreateQueue && sconfig.QueueName != "" {
+		resp, err := client.CreateQueue(&sqs.CreateQueueInput{
+			QueueName: &sconfig.QueueName,
+		})
+		if err != nil {
+			return err
+		}
+		if resp == nil || resp.QueueUrl == nil {
+			return fmt.Errorf("invalid response from AWS: missing queue URL")
+		}
+		queueURL = *resp.QueueUrl
+	} else if sconfig.QueueURL != "" {
+		queueURL = sconfig.QueueURL
+	} else {
+		resp, err := client.GetQueueUrl(&sqs.GetQueueUrlInput{
+			QueueName: &sconfig.QueueName,
+		})
+		if err != nil {
+			return err
+		}
+		if resp == nil || resp.QueueUrl == nil {
+			return fmt.Errorf("invalid response from AWS: missing queue URL")
+		}
+		queueURL = *resp.QueueUrl
+	}
+
+	connCtx, connCancel := context.WithCancel(context.Background())
+	conn := &sqsConnection{
+		client:        client,
+		config:        &sconfig,
+		queueURL:      queueURL,
+		ctx:           connCtx,
+		ctxCancelFunc: connCancel,
+	}
+	s.clientLock.Lock()
+	defer s.clientLock.Unlock()
+	if _, ok := s.connections[request.SubscriptionID]; ok {
+		s.killConnectionWithLock(request.SubscriptionID)
+	}
+	s.connections[request.SubscriptionID] = conn
+	return nil
+}
+
+func (s *sqsBackend) killConnection(subscriptionID string) {
+	s.clientLock.Lock()
+	defer s.clientLock.Unlock()
+	s.killConnectionWithLock(subscriptionID)
+}
+
+func (s *sqsBackend) killConnectionWithLock(subscriptionID string) {
+	conn := s.connections[subscriptionID]
+	conn.ctxCancelFunc()
+	delete(s.connections, subscriptionID)
+}
+
+func (s *sqsBackend) getConn(subscriptionID string) (*sqsConnection, error) {
+	s.clientLock.RLock()
+	defer s.clientLock.RUnlock()
+	conn, ok := s.connections[subscriptionID]
+	if !ok {
+		return nil, fmt.Errorf("invalid subscription_id")
+	}
+	return conn, nil
+}
+
+func (s *sqsBackend) SendSubscriptionEvent(subscriptionID string, eventJson string) error {
+	conn, err := s.getConn(subscriptionID)
+	if err != nil {
+		return err
+	}
+	backoff := conn.config.NewRetryBackoff()
+	err = backoff.Retry(func() error {
+		_, err = conn.client.SendMessage(&sqs.SendMessageInput{
+			MessageBody: &eventJson,
+			QueueUrl:    &conn.queueURL,
+		})
+		return err
+	})
+	if err != nil {
+		// refresh client and try again
+		s.killConnection(subscriptionID)
```

Comment:
I think this could panic because the function assumes the subscription ID is still in the map, but it could have been cleared by now as we weren't holding the lock while failing to send the event. I think `killConnectionWithLock` probably needs an existence check to ensure we always have a check done while the lock is held.

Reply:
Ah, good catch.